### PR TITLE
Rotate Component

### PR DESCRIPTION
**Pull request by Eyantra698Sumanto**
This Pull Request is to rotate the component in Arduino on cloud. The image is getting rotated correctly.
Errors:
1. The component comes back to original orientation while dragging.
2. The nodes, lead wires of all components and the display components of LCD and 7 segment display are not getting rotated.

### CODE REVIEW

File: ArduinoFrontend/src/app/Libs/CircuitElement.ts [hunk 1]
```diff
@@ -410,6 +410,12 @@ export abstract class CircuitElement {
     }
     this.delete();
   }
+  /**
+   * Rotates Component
+   */
```
Comment (gupta-arpit) [inline ArduinoFrontend/src/app/Libs/CircuitElement.ts:415]: create a private attribute `rotation` to store the state.

File: ArduinoFrontend/src/app/Libs/CircuitElement.ts [hunk 2]
```diff
@@ -410,6 +410,12 @@ export abstract class CircuitElement {
...
+  rotate(): void {
+    this.elements.rotate(90);
```
Comment (gupta-arpit) [inline ArduinoFrontend/src/app/Libs/CircuitElement.ts:417]:
```suggestion
    const bBox = this.elements.getBBox();
    const cx = bBox.x + bBox.width / 2;
    const cy = bBox.y + bBox.height / 2;
    this.elements.rotate(this.rotation, cx, cy);
```

File: ArduinoFrontend/src/app/Libs/CircuitElement.ts (hunk 1 above)
Comment (gupta-arpit) [reply ArduinoFrontend/src/app/Libs/CircuitElement.ts:415]: use the attribute inside `setDragListeners` to make it retain the state on drag

File: ArduinoFrontend/src/app/Libs/Workspace.ts [hunk 3]
```diff
@@ -496,6 +502,10 @@ export class Workspace {
       // paste
       Workspace.pasteComponent();
     }
+    if (event.ctrlKey && (event.key === 'a' || event.key === 'A')) {
```
Comment (gupta-arpit) [inline ArduinoFrontend/src/app/Libs/Workspace.ts:505]: This should not be needed, right?

File: ArduinoFrontend/src/app/Libs/CircuitElement.ts (hunk 2 above)
Comment (gupta-arpit) [reply ArduinoFrontend/src/app/Libs/CircuitElement.ts:417]: all the children elements should be rotated around the center of the parent. else, they will be rotated around their own centre.

Comment (gupta-arpit) [reply ArduinoFrontend/src/app/Libs/CircuitElement.ts:417]: you can probably write a separate function which returns the centreX and centreY of the element

Comment (Eyantra698Sumanto) [reply ArduinoFrontend/src/app/Libs/CircuitElement.ts:417]: Ok thanks

**Comment by firuza** [comment]
@Eyantra698Sumanto 
Overall the rotate feature seems to work but there are some issues shown below.

**1) Rotate breadboard and drag**
roate1

**2) Cannot place the component on the breadboard after rotating them**
rotate2

**3) Slider animation after rotating does not function as expected**
rotate3

**4) In this circuit, the LED should glow. It works on the other branch and on our test server. Note that no component is rotated. Is there something in your code that is breaking this? Consider this issue as the most important one.**
rotate4

**5) Issue with wires after rotating the breadboard. The wires are no longer connected to the breadboard after rotating.**
rotate5

**Comment by Eyantra698Sumanto** [comment]
Solved error 1,2 and 4

**Comment by firuza** [comment]
Thanks @Eyantra698Sumanto . 
Issue 1 is resolved. 

However, 2 and 4 are still not resolved. 

(2) I am not able to connect a component (LED) on the breadboard after rotating it. The LED gets connected to only certain points on the breadboard. 
(4) the LED does not glow.

**Pull request closed without merging**